local-app is crashing with old token

### Bug description

We need to detect it and invalidate token.

### Steps to reproduce

```
time="2021-08-25T09:40:19+02:00" level=info msg="writing workspace ssh_config file" ssh_config=/var/folders/5b/dtz20_cs0414cj898m1725pr0000gn/T/gitpod_ssh_config-89628-f8vKrNHDCwDn
time="2021-08-25T09:40:19+02:00" level=warning msg="cannot get workspace" WorkspaceID=aqua-wolverine-r61bvnkw error="jsonrpc2: code 403 message: not allowed"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1583421]

goroutine 10 [running]:
github.com/gitpod-io/local-app/pkg/bastion.(*Bastion).Update(0xc0001e1360, 0xc000332060, 0x17, 0x16a99a0)
        github.com/gitpod-io/local-app/pkg/bastion/bastion.go:259 +0x241
github.com/gitpod-io/local-app/pkg/bastion.(*LocalAppService).ResolveSSHConnection(0xc0000a4d08, 0x182f360, 0xc0003201b0, 0xc0003220f0, 0xc0000a4d08, 0xc0003201b0, 0xc00010dba0)
        github.com/gitpod-io/local-app/pkg/bastion/service.go:68 +0x49
github.com/gitpod-io/gitpod/local-app/api._LocalApp_ResolveSSHConnection_Handler(0x16b35e0, 0xc0000a4d08, 0x182f360, 0xc0003201b0, 0xc00031e1e0, 0x0, 0x182f360, 0xc0003201b0, 0xc000346080, 0x3f)
        github.com/gitpod-io/gitpod/local-app/api@v0.0.0-00010101000000-000000000000/localapp_grpc.pb.go:169 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001fefc0, 0x18355b8, 0xc000116240, 0xc000166000, 0xc000268630, 0x1bd64d8, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.39.1/server.go:1293 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0001fefc0, 0x18355b8, 0xc000116240, 0xc000166000, 0x0)
        google.golang.org/grpc@v1.39.1/server.go:1618 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000037820, 0xc0001fefc0, 0x18355b8, 0xc000116240, 0xc000166000)
        google.golang.org/grpc@v1.39.1/server.go:941 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.39.1/server.go:939 +0x1fd
```

### Expected behavior

_No response_

### Example repository

_No response_

### Anything else?

_No response_

--- a/components/gitpod-db/src/typeorm/user-db-impl.ts
+++ b/components/gitpod-db/src/typeorm/user-db-impl.ts
@@ -174,6 +174,14 @@ export class TypeORMUserDBImpl implements UserDB {
         return { user: token.user, token };
     }
 
+    public async findGitpodTokensOfUser(userId: string, tokenHash: string): Promise<GitpodToken | undefined> {
+        const repo = await this.getGitpodTokenRepo()
+        const qBuilder = repo.createQueryBuilder('gitpodToken')
+            .leftJoinAndSelect("gitpodToken.user", "user");
+        qBuilder.where('user.id = :userId AND gitpodToken.tokenHash = :tokenHash', { userId, tokenHash });
+        return qBuilder.getOne();
+    }
+
     public async findAllGitpodTokensOfUser(userId: string): Promise<GitpodToken[]> {
         const repo = await this.getGitpodTokenRepo()
         const qBuilder = repo.createQueryBuilder('gitpodToken')

--- a/components/gitpod-db/src/user-db.ts
+++ b/components/gitpod-db/src/user-db.ts
@@ -111,6 +111,7 @@ export interface UserDB extends OAuthUserRepository, OAuthTokenRepository {
     findUserByName(name: string): Promise<User | undefined>;
 
     findUserByGitpodToken(tokenHash: string, tokenType?: GitpodTokenType): Promise<{ user: User, token: GitpodToken } | undefined>;
+    findGitpodTokensOfUser(userId: string, tokenHash: string): Promise<GitpodToken | undefined>;
     findAllGitpodTokensOfUser(userId: string): Promise<GitpodToken[]>;
     storeGitpodToken(token: GitpodToken & { user: DBUser }): Promise<void>;
     deleteGitpodToken(tokenHash: string): Promise<void>;

--- a/components/gitpod-protocol/go/gitpod-service.go
+++ b/components/gitpod-protocol/go/gitpod-service.go
@@ -33,6 +33,7 @@ type APIInterface interface {
 	DeleteOwnAuthProvider(ctx context.Context, params *DeleteOwnAuthProviderParams) (err error)
 	GetBranding(ctx context.Context) (res *Branding, err error)
 	GetConfiguration(ctx context.Context) (res *Configuration, err error)
+	GetGitpodTokenScopes(ctx context.Context, tokenHash string) (res []string, err error)
 	GetToken(ctx context.Context, query *GetTokenSearchOptions) (res *Token, err error)
 	GetPortAuthenticationToken(ctx context.Context, workspaceID string) (res *Token, err error)
 	DeleteAccount(ctx context.Context) (err error)
@@ -106,6 +107,8 @@ const (
 	FunctionGetBranding FunctionName = "getBranding"
 	// FunctionGetConfiguration is the name of the getConfiguration function
 	FunctionGetConfiguration FunctionName = "getConfiguration"
+	// FunctionGetGitpodTokenScopes is the name of the GetGitpodTokenScopes function
+	FunctionGetGitpodTokenScopes FunctionName = "getGitpodTokenScopes"
 	// FunctionGetToken is the name of the getToken function
 	FunctionGetToken FunctionName = "getToken"
 	// FunctionGetPortAuthenticationToken is the name of the getPortAuthenticationToken function
@@ -495,6 +498,26 @@ func (gp *APIoverJSONRPC) GetConfiguration(ctx context.Context) (res *Configurat
 		return
 	}
 	res = &result
+
+	return
+}
+
+// GetGitpodTokenScopes calls getGitpodTokenScopes on the server
+func (gp *APIoverJSONRPC) GetGitpodTokenScopes(ctx context.Context, tokenHash string) (res []string, err error) {
+	if gp == nil {
+		err = errNotConnected
+		return
+	}
+	var _params []interface{}
+
+	_params = append(_params, tokenHash)
+
+	var result []string
+	err = gp.C.Call(ctx, "getGitpodTokenScopes", _params, &result)
+	if err != nil {
+		return
+	}
+	res = result
 
 	return
 }

--- a/components/gitpod-protocol/go/mock.go
+++ b/components/gitpod-protocol/go/mock.go
@@ -405,6 +405,21 @@ func (mr *MockAPIInterfaceMockRecorder) GetSnapshots(ctx, workspaceID interface{
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetSnapshots", reflect.TypeOf((*MockAPIInterface)(nil).GetSnapshots), ctx, workspaceID)
 }
 
+// GetGitpodTokenScopes indicates an expected call of GetGitpodTokenScopes.
+func (mr *MockAPIInterfaceMockRecorder) GetGitpodTokenScopes(ctx, tokenHash interface{}) *gomock.Call {
+	mr.mock.ctrl.T.Helper()
+	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetGitpodTokenScopes", reflect.TypeOf((*MockAPIInterface)(nil).GetGitpodTokenScopes), ctx, tokenHash)
+}
+
+// GetGitpodTokenScopes mocks base method.
+func (m *MockAPIInterface) GetGitpodTokenScopes(ctx context.Context, tokenHash string) ([]string, error) {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "GetGitpodTokenScopes", ctx, tokenHash)
+	ret0, _ := ret[0].([]string)
+	ret1, _ := ret[1].(error)
+	return ret0, ret1
+}
+
 // GetToken mocks base method.
 func (m *MockAPIInterface) GetToken(ctx context.Context, query *GetTokenSearchOptions) (*Token, error) {
 	m.ctrl.T.Helper()

--- a/components/gitpod-protocol/go/reconnecting-ws.go
+++ b/components/gitpod-protocol/go/reconnecting-ws.go
@@ -8,6 +8,7 @@ import (
 	"context"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"net/http"
 	"sync"
 	"time"
@@ -15,6 +16,19 @@ import (
 	"github.com/gorilla/websocket"
 	"github.com/sirupsen/logrus"
 )
+
+// ErrClosed is returned when the reconnecting web socket is closed.
+var ErrClosed = errors.New("reconnecting-ws: closed")
+
+// ErrBadHandshake is returned when the server response to opening handshake is
+// invalid.
+type ErrBadHandshake struct {
+	Resp *http.Response
+}
+
+func (e *ErrBadHandshake) Error() string {
+	return fmt.Sprintf("reconnecting-ws: bad handshake: code %v", e.Resp.StatusCode)
+}
 
 // The ReconnectingWebsocket represents a Reconnecting WebSocket connection.
 type ReconnectingWebsocket struct {
@@ -27,6 +41,7 @@ type ReconnectingWebsocket struct {
 	reconnectionDelayGrowFactor float64
 
 	once     sync.Once
+	closeErr error
 	closedCh chan struct{}
 	connCh   chan chan *WebsocketConnection
 	errCh    chan error
@@ -54,7 +69,12 @@ func NewReconnectingWebsocket(url string, reqHeader http.Header, log *logrus.Ent
 
 // Close closes the underlying webscoket connection.
 func (rc *ReconnectingWebsocket) Close() error {
+	return rc.closeWithError(ErrClosed)
+}
+
+func (rc *ReconnectingWebsocket) closeWithError(closeErr error) error {
 	rc.once.Do(func() {
+		rc.closeErr = closeErr
 		close(rc.closedCh)
 	})
 	return nil
@@ -69,7 +89,7 @@ func (rc *ReconnectingWebsocket) EnsureConnection(handler func(conn *WebsocketCo
 		connCh := make(chan *WebsocketConnection, 1)
 		select {
 		case <-rc.closedCh:
-			return errors.New("closed")
+			return rc.closeErr
 		case rc.connCh <- connCh:
 		}
 		conn := <-connCh
@@ -79,7 +99,7 @@ func (rc *ReconnectingWebsocket) EnsureConnection(handler func(conn *WebsocketCo
 		}
 		select {
 		case <-rc.closedCh:
-			return errors.New("closed")
+			return rc.closeErr
 		case rc.errCh <- err:
 		}
 	}
@@ -166,7 +186,7 @@ func (rc *ReconnectingWebsocket) connect(ctx context.Context) *WebsocketConnecti
 		}
 
 		dialer := websocket.Dialer{HandshakeTimeout: rc.handshakeTimeout}
-		conn, _, err := dialer.DialContext(ctx, rc.url, rc.reqHeader)
+		conn, resp, err := dialer.DialContext(ctx, rc.url, rc.reqHeader)
 		if err == nil {
 			rc.log.WithField("url", rc.url).Debug("connection was successfully established")
 			ws, err := NewWebsocketConnection(context.Background(), conn, func(staleErr error) {
@@ -177,7 +197,17 @@ func (rc *ReconnectingWebsocket) connect(ctx context.Context) *WebsocketConnecti
 			}
 		}
 
-		rc.log.WithError(err).WithField("url", rc.url).Errorf("failed to connect, trying again in %d seconds...", uint32(delay.Seconds()))
+		if err == websocket.ErrBadHandshake {
+			_ = rc.closeWithError(&ErrBadHandshake{resp})
+			return nil
+		}
+
+		var statusCode int
+		if resp != nil {
+			statusCode = resp.StatusCode
+		}
+
+		rc.log.WithError(err).WithField("statusCode", statusCode).WithField("url", rc.url).Errorf("failed to connect, trying again in %d seconds...", uint32(delay.Seconds()))
 		select {
 		case <-rc.closedCh:
 			return nil

--- a/components/gitpod-protocol/src/gitpod-service.ts
+++ b/components/gitpod-protocol/src/gitpod-service.ts
@@ -55,6 +55,7 @@ export interface GitpodServer extends JsonRpcServer<GitpodClient>, AdminServer, 
     getBranding(): Promise<Branding>;
     getConfiguration(): Promise<Configuration>;
     getToken(query: GitpodServer.GetTokenSearchOptions): Promise<Token | undefined>;
+    getGitpodTokenScopes(tokenHash: string): Promise<string[]>;
     /**
      * @deprecated
      */

--- a/components/local-app/go.mod
+++ b/components/local-app/go.mod
@@ -8,6 +8,8 @@ require (
 	github.com/gitpod-io/gitpod/gitpod-protocol v0.0.0-00010101000000-000000000000
 	github.com/gitpod-io/gitpod/local-app/api v0.0.0-00010101000000-000000000000
 	github.com/gitpod-io/gitpod/supervisor/api v0.0.0-00010101000000-000000000000
+	github.com/golang/mock v1.6.0 // indirect
+	github.com/google/go-cmp v0.5.6 // indirect
 	github.com/google/uuid v1.1.2
 	github.com/improbable-eng/grpc-web v0.14.0
 	github.com/kevinburke/ssh_config v1.1.0
@@ -30,7 +32,6 @@ require (
 	github.com/danieljoos/wincred v1.1.0 // indirect
 	github.com/fsnotify/fsnotify v1.4.9 // indirect
 	github.com/godbus/dbus/v5 v5.0.3 // indirect
-	github.com/golang/mock v1.6.0 // indirect
 	github.com/golang/protobuf v1.5.2 // indirect
 	github.com/gorilla/websocket v1.4.2 // indirect
 	github.com/grpc-ecosystem/grpc-gateway/v2 v2.5.0 // indirect

--- a/components/local-app/main.go
+++ b/components/local-app/main.go
@@ -12,7 +12,6 @@ import (
 	"time"
 
 	"context"
-	"errors"
 	"log"
 	"net/http"
 	"net/url"
@@ -93,6 +92,10 @@ func main() {
 					"GITPOD_LCA_AUTH_REDIRECT_URL",
 				},
 			},
+			&cli.BoolFlag{
+				Name:  "verbose",
+				Value: false,
+			},
 		},
 		Commands: []*cli.Command{
 			{
@@ -101,7 +104,7 @@ func main() {
 					if c.Bool("mock-keyring") {
 						keyring.MockInit()
 					}
-					return run(c.String("gitpod-host"), c.String("ssh_config"), c.Int("api-port"), c.Bool("allow-cors-from-port"), c.Bool("auto-tunnel"), c.String("auth-redirect-url"))
+					return run(c.String("gitpod-host"), c.String("ssh_config"), c.Int("api-port"), c.Bool("allow-cors-from-port"), c.Bool("auto-tunnel"), c.String("auth-redirect-url"), c.Bool("verbose"))
 				},
 				Flags: []cli.Flag{
 					&cli.PathFlag{
@@ -125,27 +128,14 @@ func DefaultCommand(name string) cli.ActionFunc {
 	}
 }
 
-func run(origin, sshConfig string, apiPort int, allowCORSFromPort bool, autoTunnel bool, authRedirectUrl string) error {
+func run(origin, sshConfig string, apiPort int, allowCORSFromPort bool, autoTunnel bool, authRedirectUrl string, verbose bool) error {
+	if verbose {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
 	logrus.WithField("ssh_config", sshConfig).Info("writing workspace ssh_config file")
 
 	// Trailing slash(es) result in connection issues, so remove them preemptively
 	origin = strings.TrimRight(origin, "/")
-	tkn, err := auth.GetToken(origin)
-	if errors.Is(err, keyring.ErrNotFound) {
-		tkn, err = auth.Login(context.Background(), auth.LoginOpts{GitpodURL: origin, RedirectURL: authRedirectUrl})
-		if tkn != "" {
-			err = auth.SetToken(origin, tkn)
-			if err != nil {
-				logrus.WithField("origin", origin).Warnf("could not write token to keyring: %s", err)
-				// Allow to continue
-				err = nil
-			}
-		}
-	}
-	if err != nil {
-		return err
-	}
-
 	originURL, err := url.Parse(origin)
 	if err != nil {
 		return err
@@ -160,32 +150,23 @@ func run(origin, sshConfig string, apiPort int, allowCORSFromPort bool, autoTunn
 		return err
 	}
 
+	var b *bastion.Bastion
+
+	client, err := connectToServer(auth.LoginOpts{GitpodURL: origin, RedirectURL: authRedirectUrl}, func() {
+		if b != nil {
+			b.FullUpdate()
+		}
+	})
+	if err != nil {
+		return err
+	}
+
 	cb := bastion.CompositeCallbacks{
 		&logCallbacks{},
 	}
 	s := &bastion.SSHConfigWritingCallback{Path: sshConfig}
 	if sshConfig != "" {
 		cb = append(cb, s)
-	}
-
-	var b *bastion.Bastion
-
-	wshost := origin
-	wshost = strings.ReplaceAll(wshost, "https://", "wss://")
-	wshost = strings.ReplaceAll(wshost, "http://", "ws://")
-	wshost += "/api/v1"
-	client, err := gitpod.ConnectToServer(wshost, gitpod.ConnectToServerOpts{
-		Context: context.Background(),
-		Token:   tkn,
-		Log:     logrus.NewEntry(logrus.New()),
-		ReconnectionHandler: func() {
-			if b != nil {
-				b.FullUpdate()
-			}
-		},
-	})
-	if err != nil {
-		return err
 	}
 
 	b = bastion.New(client, cb)
@@ -196,21 +177,100 @@ func run(origin, sshConfig string, apiPort int, allowCORSFromPort bool, autoTunn
 		// Is the origin a subdomain of the installations hostname?
 		return hostRegex.Match([]byte(origin))
 	}
-	go http.ListenAndServe("localhost:"+strconv.Itoa(apiPort), grpcweb.WrapServer(grpcServer,
-		grpcweb.WithCorsForRegisteredEndpointsOnly(false),
-		grpcweb.WithOriginFunc(allowOrigin),
-		grpcweb.WithWebsockets(true),
-		grpcweb.WithWebsocketOriginFunc(func(req *http.Request) bool {
-			origin, err := grpcweb.WebsocketRequestOrigin(req)
-			if err != nil {
-				return false
-			}
-			return allowOrigin(origin)
-		}),
-		grpcweb.WithWebsocketPingInterval(15*time.Second),
-	))
+	go func() {
+		err := http.ListenAndServe("localhost:"+strconv.Itoa(apiPort), grpcweb.WrapServer(grpcServer,
+			grpcweb.WithCorsForRegisteredEndpointsOnly(false),
+			grpcweb.WithOriginFunc(allowOrigin),
+			grpcweb.WithWebsockets(true),
+			grpcweb.WithWebsocketOriginFunc(func(req *http.Request) bool {
+				origin, err := grpcweb.WebsocketRequestOrigin(req)
+				if err != nil {
+					return false
+				}
+				return allowOrigin(origin)
+			}),
+			grpcweb.WithWebsocketPingInterval(15*time.Second),
+		))
+		if err != nil {
+			logrus.WithError(err).Error("API endpoint failed to start")
+			os.Exit(1)
+		}
+	}()
 	defer grpcServer.Stop()
 	return b.Run()
+}
+
+func connectToServer(loginOpts auth.LoginOpts, reconnectionHandler func()) (*gitpod.APIoverJSONRPC, error) {
+	tkn, err := auth.GetToken(loginOpts.GitpodURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if tkn != "" {
+		// try to connect with existing token
+		client, err := tryConnectToServer(loginOpts.GitpodURL, tkn, reconnectionHandler)
+		if client != nil {
+			return client, err
+		}
+		_, invalid := err.(*auth.ErrInvalidGitpodToken)
+		if !invalid {
+			return nil, err
+		}
+		// existing token is invalid, try again
+		logrus.WithError(err).WithField("origin", loginOpts.GitpodURL).Error()
+	}
+
+	tkn, err = login(loginOpts)
+	if err != nil {
+		return nil, err
+	}
+	return tryConnectToServer(loginOpts.GitpodURL, tkn, reconnectionHandler)
+}
+
+func tryConnectToServer(gitpodUrl string, tkn string, reconnectionHandler func()) (*gitpod.APIoverJSONRPC, error) {
+	wshost := gitpodUrl
+	wshost = strings.ReplaceAll(wshost, "https://", "wss://")
+	wshost = strings.ReplaceAll(wshost, "http://", "ws://")
+	wshost += "/api/v1"
+	client, err := gitpod.ConnectToServer(wshost, gitpod.ConnectToServerOpts{
+		Context:             context.Background(),
+		Token:               tkn,
+		Log:                 logrus.NewEntry(logrus.StandardLogger()),
+		ReconnectionHandler: reconnectionHandler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = auth.ValidateToken(client, tkn)
+	if err == nil {
+		return client, nil
+	}
+
+	closeErr := client.Close()
+	if closeErr != nil {
+		logrus.WithError(closeErr).WithField("origin", gitpodUrl).Warn("failed to close connection to gitpod server")
+	}
+
+	deleteErr := auth.DeleteToken(gitpodUrl)
+	if deleteErr != nil {
+		logrus.WithError(deleteErr).WithField("origin", gitpodUrl).Warn("failed to delete gitpod token")
+	}
+
+	return nil, err
+}
+
+func login(loginOpts auth.LoginOpts) (string, error) {
+	tkn, err := auth.Login(context.Background(), loginOpts)
+	if tkn != "" {
+		err = auth.SetToken(loginOpts.GitpodURL, tkn)
+		if err != nil {
+			logrus.WithField("origin", loginOpts.GitpodURL).Warnf("could not write token to keyring: %s", err)
+			// Allow to continue
+			err = nil
+		}
+	}
+	return tkn, err
 }
 
 type logCallbacks struct{}

--- a/components/local-app/pkg/auth/auth.go
+++ b/components/local-app/pkg/auth/auth.go
@@ -6,14 +6,18 @@ package auth
 
 import (
 	"context"
+	"crypto/sha256"
+	"encoding/hex"
 	"errors"
 	"fmt"
 	"io"
 	"net"
 	"net/http"
 	"net/url"
+	"strings"
 
 	jwt "github.com/dgrijalva/jwt-go"
+	gitpod "github.com/gitpod-io/gitpod/gitpod-protocol"
 	"github.com/sirupsen/logrus"
 	"github.com/skratchdot/open-golang/open"
 	keyring "github.com/zalando/go-keyring"
@@ -25,6 +29,49 @@ const (
 	keyringService = "gitpod-io"
 )
 
+var authScopes = []string{
+	"function:getGitpodTokenScopes",
+	"function:getWorkspace",
+	"function:getWorkspaces",
+	"function:listenForWorkspaceInstanceUpdates",
+	"resource:default",
+}
+
+type ErrInvalidGitpodToken struct {
+	cause error
+}
+
+func (e *ErrInvalidGitpodToken) Error() string {
+	return "invalid gitpod token: " + e.cause.Error()
+}
+
+// ValidateToken validates the given tkn against the given gitpod service
+func ValidateToken(client gitpod.APIInterface, tkn string) error {
+	hash := sha256.Sum256([]byte(tkn))
+	tokenHash := hex.EncodeToString(hash[:])
+	tknScopes, err := client.GetGitpodTokenScopes(context.Background(), tokenHash)
+	if e, ok := err.(*gitpod.ErrBadHandshake); ok && e.Resp.StatusCode == 401 {
+		return &ErrInvalidGitpodToken{err}
+	}
+	if err != nil && strings.Contains(err.Error(), "jsonrpc2: code 403") {
+		return &ErrInvalidGitpodToken{err}
+	}
+	if err != nil {
+		return err
+	}
+	tknScopesMap := make(map[string]struct{}, len(tknScopes))
+	for _, scope := range tknScopes {
+		tknScopesMap[scope] = struct{}{}
+	}
+	for _, scope := range authScopes {
+		_, ok := tknScopesMap[scope]
+		if !ok {
+			return &ErrInvalidGitpodToken{fmt.Errorf("%v scope is missing in %v", scope, tknScopes)}
+		}
+	}
+	return nil
+}
+
 // SetToken returns the persisted Gitpod token
 func SetToken(host, token string) error {
 	return keyring.Set(keyringService, host, token)
@@ -32,7 +79,16 @@ func SetToken(host, token string) error {
 
 // GetToken returns the persisted Gitpod token
 func GetToken(host string) (token string, err error) {
-	return keyring.Get(keyringService, host)
+	tkn, err := keyring.Get(keyringService, host)
+	if errors.Is(err, keyring.ErrNotFound) {
+		return "", nil
+	}
+	return tkn, err
+}
+
+// DeleteToken deletes the persisted Gitpod token
+func DeleteToken(host string) error {
+	return keyring.Delete(keyringService, host)
 }
 
 // LoginOpts configure the login process
@@ -123,13 +179,7 @@ func Login(ctx context.Context, opts LoginOpts) (token string, err error) {
 	conf := &oauth2.Config{
 		ClientID:     "gplctl-1.0",
 		ClientSecret: "gplctl-1.0-secret", // Required (even though it is marked as optional?!)
-		Scopes: []string{
-			"function:getWorkspace",
-			"function:getWorkspaces",
-			"function:listenForWorkspaceInstanceUpdates",
-			"resource:workspace::*::get",
-			"resource:workspaceInstance::*::get",
-		},
+		Scopes:       authScopes,
 		Endpoint: oauth2.Endpoint{
 			AuthURL:  authURL.String(),
 			TokenURL: tokenURL.String(),

--- a/components/local-app/pkg/auth/auth_test.go
+++ b/components/local-app/pkg/auth/auth_test.go
@@ -1,0 +1,86 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+//go:build linux
+// +build linux
+
+package auth
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"net/http"
+	"testing"
+
+	gitpod "github.com/gitpod-io/gitpod/gitpod-protocol"
+	"github.com/golang/mock/gomock"
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestValidateToken(t *testing.T) {
+	tkn := "foo"
+	hash := sha256.Sum256([]byte(tkn))
+	tokenHash := hex.EncodeToString(hash[:])
+
+	unauthorizedErr := &gitpod.ErrBadHandshake{
+		Resp: &http.Response{
+			StatusCode: 401,
+		},
+	}
+
+	forbiddenErr := errors.New("jsonrpc2: code 403 message: getGitpodTokenScopes")
+
+	tests := []struct {
+		Desc        string
+		Scopes      []string
+		ScopesErr   error
+		Expectation error
+	}{
+		{
+			Desc:        "invalid: unauthorized",
+			ScopesErr:   unauthorizedErr,
+			Expectation: &ErrInvalidGitpodToken{unauthorizedErr},
+		},
+		{
+			Desc:        "invalid: forbidden",
+			ScopesErr:   forbiddenErr,
+			Expectation: &ErrInvalidGitpodToken{forbiddenErr},
+		},
+		{
+			Desc:        "invalid: missing scopes",
+			Scopes:      []string{"function:getWorkspace"},
+			Expectation: &ErrInvalidGitpodToken{errors.New("function:getGitpodTokenScopes scope is missing in [function:getWorkspace]")},
+		},
+		{
+			Desc:   "valid",
+			Scopes: authScopes,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.Desc, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			gitpodAPI := gitpod.NewMockAPIInterface(ctrl)
+			gitpodAPI.EXPECT().GetGitpodTokenScopes(context.Background(), tokenHash).Times(1).Return(test.Scopes, test.ScopesErr)
+
+			var expectation string
+			if test.Expectation != nil {
+				expectation = test.Expectation.Error()
+			}
+
+			var actual string
+			err := ValidateToken(gitpodAPI, tkn)
+			if err != nil {
+				actual = err.Error()
+			}
+
+			if diff := cmp.Diff(expectation, actual); diff != "" {
+				t.Errorf("unexpected output (-want +got):\n%s", diff)
+			}
+		})
+	}
+}

--- a/components/local-app/pkg/bastion/bastion.go
+++ b/components/local-app/pkg/bastion/bastion.go
@@ -255,7 +255,8 @@ func (b *Bastion) FullUpdate() {
 func (b *Bastion) Update(workspaceID string) *Workspace {
 	ws, err := b.Client.GetWorkspace(b.ctx, workspaceID)
 	if err != nil {
-		logrus.WithError(err).WithField("WorkspaceID", workspaceID).Warn("cannot get workspace")
+		logrus.WithError(err).WithField("WorkspaceID", workspaceID).Error("cannot get workspace")
+		return nil
 	}
 	if ws.LatestInstance == nil {
 		return nil

--- a/components/server/src/auth/bearer-authenticator.ts
+++ b/components/server/src/auth/bearer-authenticator.ts
@@ -10,8 +10,6 @@ import * as crypto from 'crypto';
 import * as express from 'express';
 import { IncomingHttpHeaders } from 'http';
 import { inject, injectable } from 'inversify';
-import * as websocket from 'ws';
-import { WsNextFunction, WsRequestHandler } from '../express/ws-handler';
 import { AllAccessFunctionGuard, ExplicitFunctionAccessGuard, WithFunctionAccessGuard } from './function-access';
 import { TokenResourceGuard, WithResourceAccessGuard } from './resource-access';
 
@@ -33,7 +31,7 @@ const bearerAuthCode = 'BearerAuth';
 interface BearerAuthError extends Error {
     code: typeof bearerAuthCode
 }
-function isBearerAuthError(error: Error): error is BearerAuthError {
+export function isBearerAuthError(error: Error): error is BearerAuthError {
     return 'code' in error && error['code'] === bearerAuthCode;
 }
 function createBearerAuthError(message: string): BearerAuthError {
@@ -47,7 +45,7 @@ export class BearerAuth {
     get restHandler(): express.RequestHandler {
         return async (req, res, next) => {
             try {
-                await this.doAuth(req);
+                await this.auth(req);
             } catch (e) {
                 if (isBearerAuthError(e)) {
                     res.status(401).send(e.message);
@@ -62,7 +60,7 @@ export class BearerAuth {
     get restHandlerOptionally(): express.RequestHandler {
         return async (req, res, next) => {
             try {
-                await this.doAuth(req);
+                await this.auth(req);
             } catch (e) {
                 // don't error the request, we just have not bearer authentication token
             }
@@ -70,14 +68,7 @@ export class BearerAuth {
         }
     }
 
-    public get websocketHandler(): WsRequestHandler {
-        return async (ws: websocket, req: express.Request, next: WsNextFunction): Promise<void> => {
-            await this.doAuth(req);
-            return next();
-        }
-    }
-
-    private async doAuth(req: express.Request): Promise<void> {
+    async auth(req: express.Request): Promise<void> {
         const token = getBearerToken(req.headers)
         if (!token) {
             throw createBearerAuthError('missing Bearer token');

--- a/components/server/src/auth/rate-limiter.ts
+++ b/components/server/src/auth/rate-limiter.ts
@@ -46,6 +46,7 @@ function getConfig(config: RateLimiterConfig): RateLimiterConfig {
         "deleteOwnAuthProvider": { group: "default", points: 1 },
         "getBranding": { group: "default", points: 1 },
         "getConfiguration": { group: "default", points: 1 },
+        "getGitpodTokenScopes": { group: "default", points: 1 },
         "getToken": { group: "default", points: 1 },
         "getPortAuthenticationToken": { group: "default", points: 1 },
         "deleteAccount": { group: "default", points: 1 },

--- a/components/server/src/oauth-server/db.ts
+++ b/components/server/src/oauth-server/db.ts
@@ -5,7 +5,6 @@
  */
 
 import { OAuthClient, OAuthScope, OAuthToken } from "@jmondi/oauth2-server";
-import { ScopedResourceGuard } from "../auth/resource-access";
 
 /**
 * Currently (2021-05-15) we only support 1 client and a fixed set of scopes so hard-coding here is acceptable.
@@ -19,11 +18,11 @@ export interface InMemory {
 
 // Scopes
 const scopes: OAuthScope[] = [
+  { name: "function:getGitpodTokenScopes" },
   { name: "function:getWorkspace" },
   { name: "function:getWorkspaces" },
   { name: "function:listenForWorkspaceInstanceUpdates" },
-  { name: "resource:" + ScopedResourceGuard.marshalResourceScope({ kind: "workspace", subjectID: "*", operations: ["get"] }) },
-  { name: "resource:" + ScopedResourceGuard.marshalResourceScope({ kind: "workspaceInstance", subjectID: "*", operations: ["get"] }) }
+  { name: "resource:default" }
 ];
 
 // Clients

--- a/components/server/src/server.ts
+++ b/components/server/src/server.ts
@@ -35,7 +35,7 @@ import { WebsocketConnectionManager } from './websocket-connection-manager';
 import { DeletedEntryGC, PeriodicDbDeleter } from '@gitpod/gitpod-db/lib';
 import { OneTimeSecretServer } from './one-time-secret-server';
 import { GitpodClient, GitpodServer } from '@gitpod/gitpod-protocol';
-import { BearerAuth } from './auth/bearer-authenticator';
+import { BearerAuth, isBearerAuthError } from './auth/bearer-authenticator';
 import { HostContextProvider } from './auth/host-context-provider';
 import { CodeSyncService } from './code-sync/code-sync-service';
 import { increaseHttpRequestCounter, observeHttpRequestDuration } from './prometheus-metrics';
@@ -124,12 +124,12 @@ export class Server<C extends GitpodClient, S extends GitpodServer> {
             //  - a workspace location (ending of hostUrl.hostname)
             // We rely on the origin header being set correctly (needed by regular clients to use Gitpod:
             // CORS allows subdomains to access gitpod.io)
-            const csrfGuard: ws.VerifyClientCallbackAsync = (info: { origin: string; secure: boolean; req: http.IncomingMessage }, callback: (res: boolean, code?: number, message?: string) => void) => {
-                let allowedRequest = isAllowedWebsocketDomain(info.origin, this.env.hostUrl.url.hostname);
+            const verifyCSRF = (origin: string) => {
+                let allowedRequest = isAllowedWebsocketDomain(origin, this.env.hostUrl.url.hostname);
                 if (this.env.kubeStage === 'prodcopy' || this.env.kubeStage === 'staging') {
                     // On staging and devstaging, we want to allow Theia to be able to connect to the server from this magic port
                     // This enables debugging Theia from inside Gitpod
-                    const url = new URL(info.origin);
+                    const url = new URL(origin);
                     if (url.hostname.startsWith("13444-")) {
                         allowedRequest = true;
                     }
@@ -138,10 +138,27 @@ export class Server<C extends GitpodClient, S extends GitpodServer> {
                     log.warn("Websocket connection CSRF guard disabled");
                     allowedRequest = true;
                 }
+                return allowedRequest;
+            }
 
-                if (!allowedRequest) {
+            /**
+             * Verify the web socket handshake request.
+             */
+            const verifyClient: ws.VerifyClientCallbackAsync = async (info, callback) => {
+                if (!verifyCSRF(info.origin)) {
                     log.warn("Websocket connection attempt with non-matching Origin header: " + info.origin)
                     return callback(false, 403);
+                }
+                if (info.req.url === '/v1') {
+                    try {
+                        await this.bearerAuth.auth(info.req as express.Request)
+                    } catch (e) {
+                        if (isBearerAuthError(e)) {
+                            return callback(false, 401, e.message);
+                        }
+                        log.warn("authentication failed: ", e)
+                        return callback(false, 500);
+                    }
                 }
                 return callback(true);
             };
@@ -160,7 +177,7 @@ export class Server<C extends GitpodClient, S extends GitpodServer> {
                 }
             );
 
-            const wsHandler = new WsExpressHandler(httpServer, csrfGuard);
+            const wsHandler = new WsExpressHandler(httpServer, verifyClient);
             wsHandler.ws(websocketConnectionHandler.path, (ws, request) => {
                 const websocket = toIWebSocket(ws);
                 (request as any).wsConnection = createWebSocketConnection(websocket, console);
@@ -170,7 +187,7 @@ export class Server<C extends GitpodClient, S extends GitpodServer> {
             wsHandler.ws("/v1", (ws, request) => {
                 const websocket = toIWebSocket(ws);
                 (request as any).wsConnection = createWebSocketConnection(websocket, console);
-            }, this.bearerAuth.websocketHandler, handleError, pingPong, (ws: ws, req: express.Request) => {
+            }, handleError, pingPong, (ws: ws, req: express.Request) => {
                 websocketConnectionHandler.onConnection((req as any).wsConnection, req);
             });
         })

--- a/components/server/src/workspace/gitpod-server-impl.ts
+++ b/components/server/src/workspace/gitpod-server-impl.ts
@@ -1175,14 +1175,14 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
         const user = this.checkAndBlockUser('getHeadlessLog', { instanceId });
         const span = opentracing.globalTracer().startSpan("getHeadlessLog");
 
-        const ws = await this.workspaceDb.trace({span}).findByInstanceId(instanceId);
+        const ws = await this.workspaceDb.trace({ span }).findByInstanceId(instanceId);
         if (!ws) {
             throw new ResponseError(ErrorCodes.NOT_FOUND, `Workspace ${instanceId} not found`);
         }
 
         await this.guardAccess({ kind: 'workspaceLog', subject: ws }, 'get');
 
-        const wsi = await this.workspaceDb.trace({span}).findInstanceById(instanceId);
+        const wsi = await this.workspaceDb.trace({ span }).findInstanceById(instanceId);
         if (!wsi) {
             throw new ResponseError(ErrorCodes.NOT_FOUND, `Workspace instance for ${instanceId} not found`);
         }
@@ -1617,7 +1617,6 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
 
     public async generateNewGitpodToken(options: { name?: string, type: GitpodTokenType, scopes?: [] }): Promise<string> {
         const user = this.checkAndBlockUser("generateNewGitpodToken");
-        this.checkAndBlockUser
         const token = crypto.randomBytes(30).toString('hex');
         const tokenHash = crypto.createHash('sha256').update(token, 'utf8').digest("hex");
         const dbToken: DBGitpodToken = {
@@ -1632,6 +1631,22 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
 
         await this.userDB.storeGitpodToken(dbToken)
         return token;
+    }
+
+    public async getGitpodTokenScopes(tokenHash: string): Promise<string[]> {
+        const user = this.checkAndBlockUser("getGitpodTokenScopes");
+        let token: GitpodToken | undefined;
+        try {
+            token = await this.userDB.findGitpodTokensOfUser(user.id, tokenHash);
+        } catch (error) {
+            log.error({ userId: user.id }, "failed to resolve gitpod token: ", error);
+            return [];
+        }
+        if (!token || token.deleted) {
+            return [];
+        }
+        await this.guardAccess({ kind: "gitpodToken", subject: token }, "get");
+        return token.scopes;
     }
 
     public async deleteGitpodToken(tokenHash: string): Promise<void> {

--- a/components/server/src/workspace/workspace-starter.ts
+++ b/components/server/src/workspace/workspace-starter.ts
@@ -691,6 +691,7 @@ export class WorkspaceStarter {
             "function:storeLayout",
             "function:stopWorkspace",
             "function:getToken",
+            "function:getGitpodTokenScopes",
             "function:getContentBlobUploadUrl",
             "function:getContentBlobDownloadUrl",
             "function:accessCodeSyncStorage",

